network/src/propagation_network/mod.rs

```diff
@@ -384,6 +417,137 @@ mod test {
         }
     }
 
+    impl Network {
+        fn new() -> Self {
+            Self {
+                nodes: HashMap::new(),
+            }
+        }
+
+        /// Creates a node, but does not make it join the network.
+        fn create_node(&mut self) -> NodeKey {
+            let node = Node::new_random();
+            let key = node.id;
+            self.nodes.insert(key, node);
+            key
+        }
+
+        /// Creates n nodes.
+        fn create_nodes(&mut self, n: usize) -> Vec<NodeKey> {
+            (0..n).map(|_| self.create_node()).collect()
+        }
+
+        /// Returns nodes that have joined the network.
+        fn get_nodes_in_network(&self) -> Vec<NodeKey> {
+            self.nodes
+                .iter()
+                .filter(|(_, node)| node.network.is_some())
+                .map(|(&key, _)| key)
+                .collect()
+        }
+
+        /// Returns nodes that have joined the network.
+        fn get_nodes_not_in_network(&self) -> Vec<NodeKey> {
+            self.nodes
+                .iter()
+                .filter(|(_, node)| node.network.is_none())
+                .map(|(&key, _)| key)
+                .collect()
+        }
+
+        /// Adds all nodes to the network that haven't joined it yet.
+        async fn add_nodes_to_network_sequential(
+            &mut self,
+            max_bootstrap_points: usize,
+        ) -> Result<Vec<NodeKey>, String> {
+            let target_nodes = self.get_nodes_not_in_network();
+
+            for key in &target_nodes {
+                // Select random nodes and add them to bootstrap points.
+                let mut bootstrap_points = Vec::new();
+                let mut rng = rand::thread_rng();
```

Comment:
Can this cause a nondeterministic result of any test?
Unlike the randomness from port selection  or key generation (which wouldn't really matter), this seems quite effective.

Reply:
1. If `max_bootstrap_points == total_nodes_in_network`, the network will be initialed with fully-connected nodes, so the process and the result of peer discovery are deterministic. 
2. If we select random bootstrap nodes less than the number of total nodes in the network, the process of peer discovery will be non-deterministic. However, the expected result should always be the same if our code works well; each node figures out and keeps 20 nodes by querying given bootstrap nodes.